### PR TITLE
Extend API and websocket

### PR DESCRIPTION
**Pull request by btcontract**
Previously websocket was very basic and missed many important events. Now it sends Json-serialized notifications about sent/received/relayed payments as well as `ChannelPersisted` event so client can make timely backups.

### CODE REVIEW

**Comment by btcontract** [comment]
To summarize:

- websocket has been extended to support payment sent/relayed/received/failed events, respected messages were amended with `tag` fields so clients would know what kind of Json message they get.

- `AuditDb` can now return a `PaymentReceived` with extended payment details, most notably an actual sum paid since it may differ from the one specified in a payment request (if any), a `receivedpaymentinfo` has been added.

- `AuditDb` can now return a `PaymentSent` with extended payment details, most notably an off-chain fee paid, a `receivedpaymentinfo` has been added.

- A `pendingpayments` API call has been added to return a list of all outgoing payments which are pending right now. Combined with `receivedpaymentinfo` it can be used to determine whether an outgoing payment is sent successfully, or still being in-flight, or has been failed already.

- Once an outgoing payment is initiated a `PaymentLifecycle` now records it's inital height in `PendingPaymentDb` and on each new block a `Channel` updates a delay for outgoing payments which are still pending. A statistical method on top of that allows to find payee node ids whose payments are delayed for unusually long times.

**Comment by n1bor** [comment]
Would be nice to have the tag on all messages returned by sentinfo
1) NoSuchPayment
2) PaymentFailed
3) PaymentSucceeded
4) PaymentPending

Then can just case on the contents of tag and decide what to do.
As it is "no such payment returns" special error message - and not sure what pending does as tricky to recreate case.

**Comment by btcontract** [comment]
@n1bor 
> Would be nice to have the tag on all messages returned by sentinfo

OK, I'll add it here but in my opinion one should not rely on `send` command output, it better return nothing at all. The reason is payment may take from sub-seconds to days to complete and may even never be completed at all so someone will have to decide whether to continue having a channel with a stuck payment or break it.

Pending is simply a payment which is neither failed nor fulfilled yet. It's easy to recreate this case if you have two nodes with payee node accepting a payment but not fulfilling it.

**Comment by n1bor** [comment]
@btcontract Ignoring what send returns would be an ideal solution. But that is only possible if sentinfo returns one of the 4 above. Currently if the payment has failed it returns
`command failed: no such payment sent yet`
from which you can't tell if it was just that the payment failed or if you have a bug in your client app and never tried to pay it.

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentLifecycle.scala [hunk 1]
```diff
@@ -61,6 +62,7 @@ class PaymentLifecycle(sourceNodeId: PublicKey, router: ActorRef, register: Acto
         stop(FSM.Normal)
       } else {
         register ! Register.ForwardShortId(firstHop.lastUpdate.shortChannelId, cmd)
+        if (hops.size > 1) nodeParams.pendingPaymentDb.add(c.paymentHash, firstHop.nextNodeId, c.targetNodeId, firstHop.lastUpdate.cltvExpiryDelta, currentHeight, currentHeight, cmd.expiry)
```
Comment (n1bor) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentLifecycle.scala]: This table has slightly confused me. It never seems to delete payments when they have been paid or failed? I had assumed it only had payments that are in flight?

Comment (btcontract) [reply eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentLifecycle.scala]: Yes, they are never deleted. This table's function is to fixate a delays each payment takes and later use that info for risk analysis when evaluating if future payments to a given payee should be issued or not.

Comment (btcontract) [reply eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentLifecycle.scala]: Currently pending payments can be extracted from `channel` output, they don't need to be stored in a special table.

Comment (sstone) [reply eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentLifecycle.scala]: Yes, you can easily identify payments by channel id+ htlc id, so if you can make `sendpayment` return channel id and htlc id you can then have a `checkpayment` query that is not expensive and does not require too many internal changes.

Comment (btcontract) [reply eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentLifecycle.scala]: > you can easily identify payments by channel id+ htlc id

payment may hang in one channel for some time and then be re-sent through another channel, as far as I understand the only thing we can surely rely on is payment hash, and all channels should be queried.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala [hunk 2]
```diff
@@ -370,55 +371,59 @@ object Helpers {
...
-              remoteParams.toSelfDelay,
-              localDelayedPubkey,
-              localParams.defaultFinalScriptPubKey, feeratePerKwDelayed)
-            val sig = keyManager.sign(claimDelayed, keyManager.delayedPaymentPoint(localParams.channelKeyPath), localPerCommitmentPoint)
-            Transactions.addSigs(claimDelayed, sig)
-          })
-      }
+        delayed <- makeDelayedTx(txinfo.tx, "claim-htlc-success-delayed")
+        _ = eventStream.publish(PaymentSettlingOnChain(offChainAmount = MilliSatoshi(htlcAmounts(paymentHash)),
+          onChainAmount = delayed.tx.txOut.head.amount, paymentHash, delayed.tx.txid, "claim-htlc-delayed", isDone = false))
```
Comment (btcontract) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala]: `def claimCurrentLocalCommitTxOutputs` internal logic was changed to allow easy access to all parameters needed in `PaymentSettlingOnChain` message. 

A small additional benefit, I think, is that `success.tx -> delayed.tx` will only be produced if both transactions can make it to chain whereas previously `success` was always being published, even if related `delayed` did not have enough fees. This removes a useless transaction from blockchain.

**Comment by btcontract** [comment]
To summarize once again, the goal of recent commits is to make sure a service using Eclair can always find out what is going on with a payment using API alone.

1. **Incoming payment requests**

1.1 Payment has not arrived yet.
* `receivedinfo` API call returns nothing, payment should be considered PENDING until a payment request expires, afterwards it should be considered FAILED.

1.2 Payment has been settled off-chain.
* `receivedinfo` API call returns `PaymentReceived`, in addition to that websocket sends an identical `PaymentReceived` immediately, payment should be considered FULFILLED after this.

1.3 Payment has been received off-chain but channel got broken while payment was still in commit, either because remote peer or local service has published a commit tx.
* At first, websocket sends `PaymentReceived` immediately and `receivedinfo` API call returns `PaymentReceived`.

_Then channel breaking happens, it may happen fast when remote peer publishes their commit or they may become offline forever so eventually local service will be forced to publish a local commit._

* If off-chain payment is large enough for on-chain resolution then websocket sends `PaymentSettlingOnChain(isDone = false)` immediately and `receivedinfo` API call returns identical `PaymentSettlingOnChain(isDone = false)` instead of `PaymentReceived`. Once tx is confirmed websocket and `receivedinfo` return `PaymentSettlingOnChain(isDone = true)`.

* If off-chain payment is not large enough for on-chain resolution then websocket sends `PaymentLostOnChain` immediately, but `receivedinfo` API call keeps returning `PaymentReceived` (is this the right way?).

_How to react to this is up to service since it may be very specific._

2. **Outgoing payments**

Since payment may take a long time a general strategy should be to call `send` API endpoint but not wait for it's response and listen to websocket instead. In case if service monitoring software gets restarted they should call `channels` API endpoint and extract all currently PENDING outgoing payments.

2.1 Payment is still pending.
* websocket returns nothing, `sentinfo` API call also returns nothing, it is assumed service has called `channels` on restart so it knows which payments are pending and which do not exist at all.

2.2 Payment has been settled off-chain.
* `sentinfo` API call returns `PaymentSent`, in addition to that websocket sends an identical `PaymentSent` immediately, payment should be considered FULFILLED after this.

2.3 Payment has been failed off-chain.
* `sentinfo` API call returns `PaymentFailed`, in addition to that websocket sends an identical `PaymentFailed` immediately, payment should be considered FAILED after this.

2.4 Payment was pending for some time but then a channel got broken with payment present in commit, either because remote peer or local service has published a commit tx.
* If off-chain payment is large enough for on-chain resolution then websocket sends `PaymentSettlingOnChain(isDone = false)` immediately and `sentinfo` API call returns identical `PaymentSettlingOnChain(isDone = false)`. Payment still should be considered PENDING at this point.
  * If remote peer uses a preimage to fetch a related utxo then websocket returns `PaymentSent` immediately and `sentinfo` API call returns the same `PaymentSent` message. Payment still should be considered FULFILLED at this point.
  * If remote peer does not use a preimage then after a number of blocks we get an utxo back and then websocket returns `PaymentSettlingOnChain(isDone = true)` immediately and `sentinfo` API call returns the same `PaymentSettlingOnChain(isDone = true)` message. Payment still should be considered FAILED at this point.
* If off-chain payment is not large enough for on-chain resolution then websocket sends `PaymentLostOnChain` immediately, but `receivedinfo` API returns nothing (is this the right way?). Payment still should be considered FAILED at this point.

**Comment by sstone** [comment]
That's a hefty PR :) Thanks Anton, I'll review ASAP! You were right about HTLC identifiers, payment hash is the only reliable one we can use for pending payments

**Comment by btcontract** [comment]
Sorry for extending it all the time, but this is it, the final form (until review at least).

File: eclair-core/src/main/scala/fr/acinq/eclair/transactions/Transactions.scala [hunk 3]
```diff
@@ -50,7 +50,9 @@ object Transactions {
 
   case class CommitTx(input: InputInfo, tx: Transaction) extends TransactionWithInputInfo
   case class HtlcSuccessTx(input: InputInfo, tx: Transaction, paymentHash: BinaryData) extends TransactionWithInputInfo
-  case class HtlcTimeoutTx(input: InputInfo, tx: Transaction) extends TransactionWithInputInfo
+  case class HtlcTimeoutTx(input: InputInfo, tx: Transaction, paymentHash: BinaryData) extends TransactionWithInputInfo
+  case class HtlcTimeoutTxLegacy(input: InputInfo, tx: Transaction) extends TransactionWithInputInfo
+
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/transactions/Transactions.scala]: Why do we need both `HtlcTimeoutTxLegacy` and `HtlcTimeoutTx` ? There should be a comment in the code here and in Helpers explaining the problem and why it is implemented this way

File: eclair-core/src/test/scala/fr/acinq/eclair/wire/ChannelCodecsSpec.scala [hunk 4]
```diff
@@ -170,4 +173,27 @@ class ChannelCodecsSpec extends FunSuite {
     assert(spentMapCodec.decodeValue(spentMapCodec.encode(map).require).require === map)
   }
 
+  test("HtlcTimeoutTx compatibility") {
```
Comment (sstone) [inline eclair-core/src/test/scala/fr/acinq/eclair/wire/ChannelCodecsSpec.scala]: Same as above: compatibility issues should be explained more precisely

File: eclair-core/src/main/scala/fr/acinq/eclair/db/PendingPaymentDb.scala [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+  def addLostOnChain(paymentLostOnChain: PaymentLostOnChain)
+
+  def getSettlingOnChain(paymentHashOrTxid: BinaryData): Option[PaymentSettlingOnChain]
+
+  def getLostOnChain(paymentHash: BinaryData): Option[PaymentLostOnChain]
+
+}
+
+case class RiskInfo(targetNodeId: PublicKey, sinceBlockHeight: Long, total: Long, mean: Double, sdTimes: Double, delays: Seq[Long], adjusted: Seq[Long])
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/db/PendingPaymentDb.scala]: There should be a comment in the code explaining explaining what type of risk information is managed and how it is used

**Comment by sstone** [comment]
> Should I somehow extract those changes and put them in a separate PR?

Yes I think it should be a separate PR. The split between "legacy" and "current" tx and the addition of a payment hash field is an important change. I would like to understand the intent behind the change (beyond the compatibility issue with stored channel data), and if you think that there is a problem with the current implementation if should be discussed and split from the API extension

**Comment by btcontract** [comment]
I've decided to decompose this PR into a series of topical PR starting from https://github.com/ACINQ/eclair/pull/744 and continuing with https://github.com/ACINQ/eclair/pull/747 for now.

**Pull request closed without merging**